### ISSUE TITLE
apollograpqhl - readQuery from Store -> new references

### ISSUE THREAD

**Issue by ghost23** [apollographql/apollo-client#3814]
Hello,

In this blog post:

https://blog.apollographql.com/apollo-client-0-8-e8db5efc75fb#889f

it says:

> Since 0.7, Apollo Client returns referentially equal results when parts of a query result haven’t changed.

Is this still true for ApolloGraphQL, version 2.x ? Because for example, when I do a client.readQuery two times with the same query right one after another and then compare the results, they are not referentially equal.

Which makes implementing proper shouldComponentUpdate functions in React a bit tricky.

**Comment by CarloPalinckx** [comment]
I set up a quick test that seems to point out that this is not the case:

https://github.com/CarloPalinckx/apollo-client/blob/readquery-referential/packages/apollo-cache-inmemory/src/__tests__/cache.ts#L204

which results in failure:

```js
    expect(received).toBe(expected) // Object.is equality

    Expected: {"a": 1}
    Received: {"a": 1}
```

Not sure whether it should or shouldn't though. Maybe someone from the apollo team could chime in on this?

**Comment by hwillson** [comment]
https://github.com/apollographql/apollo-client/pull/4514 helps address this (and is available as of `apollo-client` 2.6.0). Thanks!
